### PR TITLE
Add ability to white list IP addresses to ignore iprestrictions

### PR DESCRIPTION
**Pull request by andrewhancox**
This is an enhancement a client required:

The Wordpress site requesting the tokens was running on a server in the clients DMZ and the Moodle was hosted externally, the meant that for internal users the IP detected by Wordpress and to which the login token is locked is on an internal range so when the user then reaches Moodle via a public IP it looks like something dodgy is going on.
This setting allows you to whitelist private address ranges etc. so that you can keep the IP restrictions in place for external users.

Now I've written all that I guess it's a bit of an edge case... I wrote unit tests though...

### CODE REVIEW

File: auth.php [hunk 1]
```diff
@@ -603,14 +605,14 @@ protected function get_user_fields_parameters() {
 
         $mappingfield = $this->get_mapping_field();
         if ($this->should_create_user() || $this->should_update_user()) {
-            $parameters['firstname'] = new external_value(PARAM_NOTAGS, 'The first name(s) of the user', VALUE_OPTIONAL);
-            $parameters['lastname']  = new external_value(PARAM_NOTAGS, 'The family name of the user', VALUE_OPTIONAL);
+            $parameters['firstname'] = new external_value(core_user::get_property_type('firstname'), 'The first name(s) of the user', VALUE_OPTIONAL);
```
Comment (dmitriim) [inline auth.php]: Looks like this function is for 3.1+

File: classes/core_userkey_manager.php [hunk 2]
```diff
@@ -124,8 +124,22 @@ public function validate_key($keyvalue) {
...
+                $whitelisted = false;
+                foreach ($ips as $ip) {
+                    if (address_in_subnet($remoteaddr, $ip)) {
+                        $whitelisted = true;
+                    }
+                }
+                if (!$whitelisted) {
+                    print_error('ipmismatch', 'error', '', null, "Remote address: $remoteaddr\nKey IP: $key->iprestriction");
+                }
+            } else if (empty($remoteaddr) ) {
```
Comment (dmitriim) [inline classes/core_userkey_manager.php]: Should we check this first and throw an error if empty?

File: tests/auth_plugin_test.php [hunk 3]
```diff
@@ -843,6 +844,84 @@ public function test_ipmismatch_exception_thrown_if_ip_is_incorrect() {
     }
 
     /**
+     * Test that IP address mismatch exception gets thrown if incorrect IP.
```
Comment (dmitriim) [inline tests/auth_plugin_test.php]: Is it correct test description?

File: tests/auth_plugin_test.php [hunk 4]
```diff
@@ -843,6 +844,84 @@ public function test_ipmismatch_exception_thrown_if_ip_is_incorrect() {
...
+        $key->timecreated   = time();
+        $DB->insert_record('user_private_key', $key);
+
+        $_POST['key'] = 'IpmismatchKey';
+        $_SERVER['HTTP_CLIENT_IP'] = '192.168.1.2';
+        @$this->auth->user_login_userkey();
+    }
+
+    /**
+     * Test that IP address mismatch exception gets thrown if incorrect IP.
```
Comment (dmitriim) [inline tests/auth_plugin_test.php]: Is it correct test description?

File: tests/auth_plugin_test.php [hunk 5]
```diff
@@ -843,6 +844,84 @@ public function test_ipmismatch_exception_thrown_if_ip_is_incorrect() {
...
+     *
+     * @expectedException moodle_exception
+     * @expectedExceptionMessage Unsupported redirect to http://www.example.com/moodle detected, execution terminated.
+     */
+    public function test_ipmismatch_exception_notthrown_if_ip_is_whitelisted_temp() {
```
Comment (dmitriim) [inline tests/auth_plugin_test.php]: temp? is it wrong function name?

File: auth.php (hunk 1 above)
Comment (andrewhancox) [reply auth.php]: This shouldn't be in there - I fudged the merge from the Moodle code base into this plugin repo. I've taken these changes out.

File: classes/core_userkey_manager.php (hunk 2 above)
Comment (andrewhancox) [reply classes/core_userkey_manager.php]: Probably... Have changed to check empty($remoteaddr) first

File: tests/auth_plugin_test.php (hunk 3 above)
Comment (andrewhancox) [reply tests/auth_plugin_test.php]: No, was copy and pasted and not changed, fixed.

File: tests/auth_plugin_test.php (hunk 4 above)
Comment (andrewhancox) [reply tests/auth_plugin_test.php]: No, was copy and pasted and not changed, fixed.

File: tests/auth_plugin_test.php (hunk 5 above)
Comment (andrewhancox) [reply tests/auth_plugin_test.php]: This was a very client specific test that I didn't intend to commit. I have removed it. Tried to do things to quickly...

**Comment by andrewhancox** [comment]
I've dealt with the issues from your review.

**Comment by dmitriim** [comment]
Hey @andrewhancox! Some unit test are failing on php5.6 and php7 Check out travis report please https://travis-ci.org/catalyst/moodle-auth_userkey/builds/282826163?utm_source=github_status&utm_medium=notification

**Comment by andrewhancox** [comment]
The issue was with how I was detecting successful logins in my unit tests, not the PHP version. I have refactored to test in a way that will work across all PHP versions and they should now all pass... The one that is failing appears to be related to a core library in the master branch of Moodle now require PHP 7
